### ISSUE TITLE
API error response body JSON not consistent with Docker convention

### ISSUE THREAD

**Issue by AngieCris** [vmware/harbor#3124]
Upon unexpected HTTP status code, the error response body Harbor API returns has the following JSON structure:
```
{ 
  "code: <error code>,
  "message": <error message>,
  "detail": 
}
```
However, the error for unexpected HTTP status code from the Docker API has this JSON structure (documentation: https://docs.docker.com/registry/spec/api/#base)
 ```
{
	"errors:" [
	    {
            "code": <error code>,
            "message": "<error message>",
            "detail": ...
            },
        ...
    ]
}
```

Due to this inconsistency, VIC cannot extract the `message` field and provide useful information to the user (only a generic error message is returned to the client). See: https://github.com/vmware/vic/issues/5951

For example, when an image violates the project level content trust, a docker pull from Harbor will result in this error response:
```
{"code":412,"message":"The image is not signed in Notary.","details":"The image is not signed in Notary."}
```
But the message that client gets from VCH will be the following:
```
Unexpected http code: 412 (Precondition Failed), URL: https://......
```
The client side can't see the detailed error message on the command line output, due to the reason described above. 
In the VCH docker-personality debug log, the client can see the complete error response body in string. This is the message in the debug log:
```
Aug 25 2017 16:50:44.215Z DEBUG Error message json string: {"code":412,"message":"The image is not signed in Notary.","details":"The image is not signed in Notary."}
```

It would be better if the error response format can match Docker API, so client can get more useful information. The release note could say that if a generic error message is returned, check the docker-personality server debug log for more details.

**Comment by mdubya66** [comment]
The engine issue is closed so adding kind/note to this. https://github.com/vmware/vic/issues/5951

**Comment by wy65701436** [comment]
this kind of error message is compatible and it's readable for docker client, see https://github.com/moby/moby/blob/master/vendor/github.com/docker/distribution/registry/client/errors.go#L48.

@reasonerjt, we can discuss whether to update return error format.

**Comment by AngieCris** [comment]
If I got it right (and correct me if I'm wrong) - Docker accepts both irregularly formatted err json with a `details` field, and err json with the same format as Docker. Thus Harbor can have different err json format but still be compatible with Docker client.
In the code from the link in @wy65701436 comment:
line 53 https://github.com/moby/moby/blob/master/vendor/github.com/docker/distribution/registry/client/errors.go#L53 unmarshals the error json and sees if it contains a `details` field. Since Harbor err json has a `details` field, this is compatible. 
line 65 https://github.com/moby/moby/blob/master/vendor/github.com/docker/distribution/registry/client/errors.go#L65 the unmarshals the err json into the same format as Docker (in https://github.com/docker/distribution/blob/master/registry/api/errcode/errors.go).

In the err code Docker provides (https://github.com/docker/distribution/blob/master/registry/api/errcode/errors.go), it wraps all errors in an array.

**Comment by sflxn** [comment]
It's a simple enough fix for either the engine or harbor to do.  We should do what's expedient for now.  However, Harbor should be aware that what they're doing is conforming to an old format that may get phased out at any moment by Docker.  They make no guarantees with regards to full backwards compatibility, as seen from how they handle API versions.  Sooner or later, Harbor will likely need to update the error output to the updated version.

**Comment by reasonerjt** [comment]
Fixed in #3125
